lib/collection/src/collection/clean.rs

```diff
@@ -0,0 +1,318 @@
+            .collect()
+    }
+}
+
+/// A background task for cleaning a shard
+///
+/// The task will delete points that don't belong in the shard according to the hash ring. This is
+/// used in context of resharding, where points are transferred to different shards.
+pub(super) struct ShardCleanTask {
+    /// Handle of the clean task
+    #[allow(dead_code)]
+    handle: JoinHandle<()>,
+    /// Watch channel with current status of the task
+    status: Receiver<ShardCleanStatus>,
+    /// Cancellation token drop guard, cancels the task if this is dropped
+    cancel: DropGuard,
+}
+
+impl ShardCleanTask {
+    /// Create a new shard clean task and immediately execute it
+    pub fn new(shards_holder: &Arc<LockedShardHolder>, shard_id: ShardId) -> Self {
+        let (sender, receiver) = tokio::sync::watch::channel(ShardCleanStatus::Started);
+        let shard_holder = Arc::downgrade(shards_holder);
+        let cancel = CancellationToken::default();
+
+        let task = tokio::task::spawn(Self::task(sender, shard_holder, shard_id, cancel.clone()));
+
+        ShardCleanTask {
+            handle: task,
+            status: receiver.clone(),
+            cancel: cancel.drop_guard(),
+        }
+    }
+
+    pub fn is_cancelled_or_failed(&self) -> bool {
+        matches!(
+            self.status.borrow().deref(),
+            ShardCleanStatus::Cancelled | ShardCleanStatus::Failed(_)
+        )
+    }
+
+    pub fn abort(self) {
+        // Explicitly cancel clean task
+        // We don't have to because the drop guard does it for us, but this makes it more explicit
+        self.cancel.disarm().cancel();
+    }
+
+    async fn task(
+        sender: Sender<ShardCleanStatus>,
+        shard_holder: Weak<LockedShardHolder>,
+        shard_id: ShardId,
+        cancel: CancellationToken,
+    ) {
+        let mut offset = None;
+
+        let status = loop {
+            // Check if cancelled
+            if cancel.is_cancelled() {
+                break ShardCleanStatus::Cancelled;
+            }
+
+            // Get shard
+            let Some(shard_holder) = shard_holder.upgrade() else {
+                break ShardCleanStatus::Failed("Shard holder dropped".into());
+            };
+            let shard_holder = shard_holder.read().await;
+            let Some(shard) = shard_holder.get_shard(shard_id) else {
+                break ShardCleanStatus::Failed(format!("Shard {shard_id} not found"));
+            };
+            if !shard.is_local().await {
+                break ShardCleanStatus::Failed(format!("Shard {shard_id} is not a local shard"));
+            }
+
+            // Scroll batch of points with hash ring filter
+            let filter = shard_holder
+                .hash_ring_filter(shard_id)
+                .expect("hash ring filter");
+            let filter = Filter::new_must_not(Condition::CustomIdChecker(Arc::new(filter)));
+            let mut ids = match shard
```

Comment:
can you please add a `TODO(ratelimiter) this scroll should not be rate limited` for me :)